### DISCUSSION TITLE
Supporting WebAuthn Conditional UI

### DISCUSSION THREAD

**Discussion by wadahiro** [keycloak/keycloak#15860] [Ideas]
Hello,

I would like to propose support for [WebAuthn Conditional UI](https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI) in Keycloak.

# Motivation

[WebAuthn Conditional UI](https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI) is currently supported in Chrome[^1] and Safari[^2]. It brings a better login experience using WebAuthn/FIDO2. ID-less authentication using WebAuthn/FIDO2 can be integrated into the normal login ID/password input form. This allows both users using password authentication and those using WebAuthn to use the same login screen. Users using WebAuthn can then log in with fewer operations.

The following GIF video shows the login operation of a prototype implementation of the WebAuthn Conditional UI in Keycloak. First, the login is performed using WebAuthn. In the login ID input field, the browser's autofill automatically pops up a pre-registered Passkey (security key). The user simply selects it and the WebAuthn login is triggered.
In the video, the user then logs in using the normal login ID and password. In this way, the exact same login form can be used to provide traditional password-based login and WebAuthn login screen.

![keyclaok-conditional-ui-poc](https://user-images.githubusercontent.com/28739/205900001-aa814dfc-31ed-43a0-a1d9-2a9d7e7559e7.gif)

I believe that Keycloak's native support for this will provide a better UX for end-users.

[^1]: https://developer.chrome.com/blog/webauthn-conditional-ui/
[^2]: https://webkit.org/blog/13152/webkit-features-in-safari-16-0/ 

# Proposal

I have considered the following two approaches to implementing WebAuthn Conditional UI in Keycloak. I would like to get your feedback.

## Proposal 1

Keycloak's normal login screen uses the [UsernamePasswordForm](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/UsernamePasswordForm.java) or [UsernameForm](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/UsernameForm.java) and [PasswordForm](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/PasswordForm.java) when using the standard authenticators. First approach is implementing it into these existing Authenticators.

This means that these Authenticators will be extended to handle WebAuthn authentication request.
When a WebAuthn authentication request comes in, the functionality needs to be extended to delegate processing to the [WebAuthnPasswordlessAuthenticator](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticator.java).

The Freemaker templates of these authenticators also need to be modified to support WebAuthn Conditional UI.

## Proposal 2

Add new Authenticator(s) with WebAuthn Conditional UI implementation. This means that no changes are made to the existing UsernamePasswordForm, UsernameForm and PasswordForm. This approach does not affect existing functionality, but we will have duplicated code.

**Comment by denniskniep** [comment]
I would like to propose another slightly modified version of "Proposal 1"

### Proposal 1a
 The [UsernamePasswordForm](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/UsernamePasswordForm.java) , [UsernameForm](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/UsernameForm.java) and [PasswordForm](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/PasswordForm.java)  will be extended to handle webAuthN requests and delegate processing to [WebAuthnPasswordlessAuthenticator](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticator.java) (so far same as Proposal 1)

Additionally a config is added to the Authenticators containing a feature flag which can be activated to support Conditional UI feature for passwordless authentication. This feature flag is by default off. Therefore we are not affecting existing functionality and do not duplicate code

**Comment by jonkoops** [comment]
I certainly think this would be a great enhancement, and I would agree that proposal 1 is the best way to go. Feel free to log an issue, so we can prioritize this as a feature.

**Reply by embesozzi** [reply]
In relation to supporting [WebAuthn Conditional UI](https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI), here's an custom authenticator [keycloak-webauthn-conditional-mediation](https://github.com/embesozzi/keycloak-webauthn-conditional-mediation)  that provides this support.  It extends the existing [WebAuthnPasswordlessAuthenticator](https://github.com/embesozzi/keycloak-webauthn-conditional-mediation/blob/main/src/main/java/com/twogenidentity/keycloak/authenticator/WebAuthnConditionalMediationAuthenticator.java#L14) to include WebAuthn Conditional UI functionality. Perharps it'll help as a reference.

**Comment by dasniko** [comment]
Great discussion! 🎉
I also implemented a first version of a WebAuthn Conditional UI authenticator. 😊

My approach was to extend the default `WebAuthnPasswordlessAuthenticator` and delegating the work to the `UsernamePasswordForm` if a username has been sent (could improve this check, still it's only on demo level, but works).
With this approach, I don't need duplicated code, my `ConditionalWebAuthnPasswordlessAuthenticator` is just a kind of a facade in front of the two default available authenticators.

A feature which I think is pretty useful, is to know, through which method the user authenticated, so that it'd become possible to distinguish during the auth flow, if a user "only" used username/password, to also enforce e.g. OTP as a 2nd factor. Therefor, I write the authenticator used to the AuthSessionNotes and implemented a custom conditional authenticator which checks theses AuthSessionNotes.

Authenticator Code:
* https://github.com/dasniko/keycloak-extensions-demo/tree/main/passkey/src/main/java/dasniko/keycloak/passkey/authentication
* https://github.com/dasniko/keycloak-extensions-demo/blob/main/passkey/src/main/resources/theme-resources/templates/webauthn-conditional-authenticate.ftl

Conditional Authenticator:
* https://github.com/dasniko/keycloak-extensions-demo/blob/main/conditional-authenticators/src/main/java/dasniko/keycloak/authentication/conditional/ConditionalAuthNoteAuthenticator.java

**Reply by embesozzi** [reply]
I like the idea about using the `AuthSessionNotes` feature to provide flexibility in customizing the authentication flow. 
In the future, when keycloak supports the use of the `amr` claim, it will helpful to return the reference of the authentication mechanism used to the application.
Perhaps, instead of using the session note name ["credentialTypeUsed"](https://github.com/dasniko/keycloak-extensions-demo/blob/main/passkey/src/main/java/dasniko/keycloak/passkey/authentication/WebAuthnConditionalAuthenticator.java#L47), we can use the name "amr" with the value following the [RFC 8176 section 2](https://www.rfc-editor.org/rfc/rfc8176.html#section-2).

**Reply by dasniko** [reply]
Thanks for your comment and thoughts about the `amr` claim.
I agree with you that "credentialTypeUsed" is not a good name. It's just a quick implementation for demo purposes, hacked in some hours... I will change that in the next iteration. :) 
But I'm not sure about the values auf the `amr` claim, as mentioned in the RFC. When using username/password, then `pwd` might be suitable, but what would you use in case of webauthn/passkey usage? Additionally, when password was used, I want to use this information to add some 2nd factor on top, then the result would have to be `otp` for example. Each authenticator would have to implement this, that it would be holistic usable.
However, it's worth every discussion!

**Reply by dasniko** [reply]
Ok, I should have read it entirely... The value "user" might be suitable for webauthn.

**Comment by oculos** [comment]
Just one idea: I like the idea of modifying the Username Password form so that a user could login right away. But it would be also nice to be able to add a button/link there to configure a Passkey for the user. This way, the authentication flow would move to a username/password only form, and then the user would redirect to create his passkey.

Correct if I am wrong, but today, one can only create a passkey if the account console is available or if the WebAuthn authentication is a default required action,  none if which are ideal in many scenarios, especially those with user federation.

**Reply by dasniko** [reply]
I'm currently implementing for a customer an overwrite for the default `WebAuthnPasswordlessRegister`, which registers itself if the user doesn't have a Passkey registered. Then, the user will be notified after he/she authenticates, that we now support Passkeys, but unlike the default required actions, the user will not be forced to register a Passkey, but is able to skip this step. If the user wants to register a Passkey later, he/she can do it from the account-console, or, like in our case, from a custom account page in our application.

**Reply by embesozzi** [reply]
Hi @oculos,
Based on your last question, the required action is not the only option nowadays. The [WebAuthnAuthenticator](https://github.com/keycloak/keycloak/blob/64836680d7f3bd2ac48962564eb3f25f82e28e91/services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnAuthenticator.java) also triggers the registration.

Another cool option is performing a progressive passkey enrollment during the sign-in process; this requires minor customization. You can experiment with this in the [keycloak-workshop-stepup-mfa-biometrics](https://github.com/embesozzi/keycloak-workshop-stepup-mfa-biometrics).

**Comment by ptr1120** [comment]
@dasniko I tested your custom authenticator [keycloak-webauthn-conditional-mediation](https://github.com/embesozzi/keycloak-webauthn-conditional-mediation). For me it provides a great user experience for login using passkeys. What about contributing the Authenticator in the keycloak repo?

**Reply by dasniko** [reply]
Your linked project is not mine, it's from @embesozzi ...

**Reply by embesozzi** [reply]
Hi @ptr1120, nowadays, passkeys are becoming a new standard, meaning that they are still in adoption. Therefore, based on my experience, you likely ended up with customizations based on customer requirements.

**Comment by FlxPeters** [comment]
Hey folks, this feature is exactly what I was looking for.

Has been a while since the last comment. Are there any plans to have this as a native feature or is the extension still the way to go?
